packages/schematics/src/app/rules.ts

```diff
@@ -0,0 +1,207 @@
+        return tree;
+      }),
+    );
+  };
+}
+
+export function installDependencies(): Rule {
+  return (tree: Tree, context: SchematicContext) => {
+    context.addTask(new NodePackageInstallTask());
+    context.logger.info('✅️ Dependencies installed');
+    return tree;
+  };
+}
+
+export function addAuguryMain(options: Schema): Rule {
+  return (tree: Tree, context: SchematicContext) => {
+    const workspace = getWorkspace(tree);
+    const project = getProject(workspace, options.project);
+
+    if (!project || !project.sourceRoot) {
+      throw new SchematicsException(`Project root could not be found`);
+    }
+
+    const sourceDir = normalize(project.sourceRoot);
+    const templSource = apply(url('./files'), [
+      template({
+        ...options,
+      }),
+      move(sourceDir),
+    ]);
+
+    context.logger.info(`✅️ Adding main.augury.ts...`);
+
+    return branchAndMerge(mergeWith(templSource))(tree, context);
+  };
+}
+
+export function addBuildConfiguration(options: Schema): Rule {
+  return (tree: Tree, context: SchematicContext) => {
+    const workspace = getWorkspace(tree);
+    const project = getProject(workspace, options.project);
+    const projectTargets = getProjectTargets(project);
+
+    if (
+      !projectTargets.build ||
+      !projectTargets.build.configurations ||
+      !projectTargets.serve ||
+      !projectTargets.serve.configurations
+    ) {
+      throw new SchematicsException("Project build configuration couldn't be found.");
+    }
+
+    projectTargets.build.configurations = {
+      ...projectTargets.build.configurations,
+      augury: {
+        fileReplacements: [
+          {
+            replace: 'src/main.ts',
+            with: 'src/main.augury.ts',
+          },
+        ],
+      },
+    };
+
+    projectTargets.serve.configurations = {
+      ...projectTargets.serve.configurations,
+      augury: {
+        browserTarget: `${options.project}:build:augury`,
+      },
+    };
+
+    context.logger.info(`✅️ Added augury build configuration to angular.json`);
+
+    return updateWorkspace(workspace)(tree, context) as Tree;
+  };
+}
+
+export function addAuguryScriptToPackage(options: Schema): Rule {
+  return (tree: Tree, context: SchematicContext) => {
+    const workspace = getWorkspace(tree);
+    const project = getProject(workspace, options.project);
+
+    if (!project || !project.sourceRoot) {
+      throw new SchematicsException(`Project root could not be found`);
+    }
+
+    const packageJsonAst = _readPackageJson(tree);
+    const scriptsNode = findPropertyInAstObject(packageJsonAst, 'scripts');
+
+    const auguryScript = {
+      name: 'start:augury',
+      command: 'ng serve --configuration augury',
+    };
+
+    const recorder = tree.beginUpdate('/package.json');
+    if (!scriptsNode) {
+      // Haven't found the script key, add it to the root of the package.json.
+      appendPropertyInAstObject(
+        recorder,
```

Comment:
One thing I found really help was to put all of the ast code in its own package. It gets large/messy at times, and when committing to refactor it, can take up a lot of code on its own. I think it's best to separate concerns. Consider `src/app/ast` folder for this.

My general folder structure was something like

```
/ast (related operations in 1 file, used several files for different tasks)
/rules (one rule per file)
/schematics
/types (I put interfaces and functions that operate specifically on those interfaces here. Schema interface and related functions can go here, for example.)
```

Reply:
I like this structure, I'll refactor the code to follow this structure.